jiant/tasks/tasks.py

```diff
@@ -3731,3 +3732,113 @@ def get_metrics(self, reset=False):
+        this function and all functions that call this function because
+        the step of reading in the data for SOP is different than other
+        PairClassificationTasks.
+
+        Args:
+            path: (str) data file path
+        """
+        f = open(path, "r")
+        #  The dataset comes with one sentence per line, thus we split by
+        #  line here.
+        text = [line.strip() for line in f]
+        for i in range(len(text) - 1):
+            if random.uniform(0, 1) > 0.5:
+                is_right_order = 1
+                sent_a = text[i]
+                sent_b = text[i + 1]
+            else:
+                is_right_order = 0
+                sent_a = text[i + 1]
+                sent_b = text[i]
+            sent_a_processed = tokenize_and_truncate(
+                self._tokenizer_name, sent_a, self.max_seq_len // 2
+            )
+            sent_b_processed = tokenize_and_truncate(
+                self._tokenizer_name, sent_b, self.max_seq_len // 2
+            )
+            yield (sent_a_processed, sent_b_processed, is_right_order)
+
+    def load_data(self):
+        pass
+
+    def process_split(
+        self, split, indexers, model_preprocessing_interface
+    ) -> Iterable[Type[Instance]]:
+        """Process a sentence order prediction split by indexing and creating fields.
+        We override the PairClassificationTask process_split because our data split
+        is different from the typical PairClassificationTask due to the more memory-efficient
+        generator way of loading data we employ for SOP due to the dataset size.
+        Args:
+            split: (list) a single list of sentences
+            indexers: (Indexer object) indexer to index input words
+        """
+
+        def _make_instance(sent_):
+            sent_a, sent_b, is_right_order = sent_
+            inp = model_preprocessing_interface.boundary_token_fn(sent_a, sent_b)
+            input_sent = sentence_to_text_field(inp, indexers)
+            label = LabelField(is_right_order, label_namespace="labels", skip_indexing=True)
+            d = {"inputs": input_sent, "labels": label}
+            return Instance(d)
+
+        for sent in split:
+            yield _make_instance(sent)
+
+    def get_split_text(self, split: str):
+        """Get split text as iterable of records.
+        Args:
+            split: (str) should be one of 'train', 'val', or 'test'.
+        """
+        return self.get_data_iter(self.files_by_split[split])
+
+    def get_sentences(self) -> Iterable[Sequence[str]]:
+        """Yield sentences, used to compute vocabulary.
+        """
+        for split in self.files_by_split:
+            # Don't use test set for vocab building.
+            if split.startswith("test"):
+                continue
+            for sent in self.get_data_iter(self.files_by_split[split]):
+                yield sent[0]
+                yield sent[1]
```

Comment:
If I'm reading this correctly, use something like
```python
for sent_a, sent_b, is_right_order in self.get_data_iter(self.files_by_split[split]):
    yield sent_a
    yield_sent_b
```
(Also it looks like almost every sentences will be yielded twice, right? This would increase memory/storage. Arguably, for tasks like MLM and NSP/SOP, we don't necessarily care about making full use of every sentence since raw text is so plentiful.)